adapters/Unity/UnityAdapter/GADMAdapterUnity.m

```diff
@@ -123,68 +130,66 @@ - (NSString *)getPlacementID {
+    } else if (newState == kUnityAdsPlacementStateNotAvailable) {
+      NSError *errorWithDescription = GADUnityErrorWithDescription(@"PlACEMENT_NOTAVAILABLE");
+      [strongNetworkConnector adapter:self didFailAd:errorWithDescription];
+    }
+  }
 }
 
 - (void)unityAdsDidFinish:(NSString *)placementID withFinishState:(UnityAdsFinishState)state {
   id<GADMAdNetworkConnector> strongNetworkConnector = _networkConnector;
-  if (strongNetworkConnector) {
-    [strongNetworkConnector adapterWillDismissInterstitial:self];
-    [strongNetworkConnector adapterDidDismissInterstitial:self];
+  if (strongNetworkConnector && [placementID isEqualToString:_placementID]) {
+    if (state == kUnityAdsFinishStateCompleted) {
+      [strongNetworkConnector adapterDidDismissInterstitial:self];
+    } else if (state == kUnityAdsFinishStateError) {
+      [strongNetworkConnector adapterWillPresentInterstitial:self];
+      [strongNetworkConnector adapterDidDismissInterstitial:self];
+    } else if (state == kUnityAdsFinishStateSkipped) {
+      [strongNetworkConnector adapterDidDismissInterstitial:self];
+    }
+    [UnityAds removeDelegate:self];
   }
 }
 
 - (void)unityAdsReady:(NSString *)placementID {
   id<GADMAdNetworkConnector> strongNetworkConnector = _networkConnector;
-  if (!_isLoading) {
-    return;
-  }
-
-  if (strongNetworkConnector) {
+  if (strongNetworkConnector && [placementID isEqualToString:_placementID]) {
     [strongNetworkConnector adapterDidReceiveInterstitial:self];
   }
-  _isLoading = NO;
 }
 
 - (void)unityAdsDidClick:(NSString *)placementID {
   id<GADMAdNetworkConnector> strongNetworkConnector = _networkConnector;
   // The Unity Ads SDK doesn't provide an event for leaving the application, so the adapter assumes
   // that a click event indicates the user is leaving the application for a browser or deeplink, and
   // notifies the Google Mobile Ads SDK accordingly.
-  if (strongNetworkConnector) {
+  if (strongNetworkConnector && [placementID isEqualToString:_placementID]) {
     [strongNetworkConnector adapterDidGetAdClick:self];
     [strongNetworkConnector adapterWillLeaveApplication:self];
   }
 }
 
 - (void)unityAdsDidError:(UnityAdsError)error withMessage:(NSString *)message {
-  id<GADMAdNetworkConnector> strongNetworkConnector = _networkConnector;
-  if (!_isLoading) {
-    // Unity Ads show error will only happen after the ad has been loaded. So, we will send
-    // dismiss/close callbacks.
-    if (error == kUnityAdsErrorShowError) {
-      if (strongNetworkConnector) {
-        [strongNetworkConnector adapterWillDismissInterstitial:self];
-        [strongNetworkConnector adapterDidDismissInterstitial:self];
-      }
-    }
-    return;
-  }
-
-  NSError *errorWithDescription = GADUnityErrorWithDescription(message);
-  if (strongNetworkConnector) {
-    [strongNetworkConnector adapter:self didFailAd:errorWithDescription];
-  }
-  _isLoading = NO;
+  //do nothing.
```

Comment:
Can you confirm that it is safer to rely on `unityAdsPlacementStateChanged` to catch errors vs catching them on `unityAdsDidError` ?